### ISSUE TITLE
[REVIEW] ai-data-privacy: add provider routing and prompt-cache residency gates

### ISSUE THREAD

**Issue by CDEN190** [UnitOneAI/SecuritySkills#2246]
## Skill Being Reviewed
**Skill name:** `ai-data-privacy`
**Skill path:** `skills/ai-security/ai-data-privacy/`

## False Positive Analysis

**Benign code that triggers a false positive:**
```ts
import { AzureOpenAI } from "openai";

const client = new AzureOpenAI({
  endpoint: process.env.AZURE_OPENAI_EU_ENDPOINT,
  apiVersion: "2025-01-01-preview",
  azureADTokenProvider,
});

const redactedPrompt = await redactPii(userPrompt);

await client.chat.completions.create({
  model: "eu-support-assistant",
  messages: [{ role: "user", content: redactedPrompt }],
  user: tenantScopedUserId,
});
```

**Why this is a false positive:**

The current skill correctly treats third-party LLM APIs as a privacy boundary, but it does not give reviewers a positive evidence gate for a provider path that is contract-bound, region-bound, redacted before transmission, and configured not to train on customer data. A reviewer can see `openai|azure.openai|api.key` and report "PII sent to a third-party LLM API" without distinguishing a compliant regional processor path from an unapproved public fallback.

The issue is not that external model APIs are always safe. The gap is that the skill lacks a structured way to prove where the data is routed, whether fallback providers preserve the same privacy commitments, and whether secondary AI infrastructure stores prompt/completion copies outside the approved boundary.

## Coverage Gaps

**Missed variant 1: provider fallback silently crosses the approved region or processor boundary**
```ts
const primary = new AzureOpenAI({
  endpoint: process.env.AZURE_OPENAI_EU_ENDPOINT,
});

const fallback = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

export async function answer(messages: ChatMessage[]) {
  try {
    return await primary.chat.completions.create({
      model: "eu-support-assistant",
      messages,
    });
  } catch {
    return await fallback.chat.completions.create({
      model: "gpt-4.1-mini",
      messages,
    });
  }
}
```

**Why it should be caught:**

The primary path may satisfy DPA, data residency, and retention requirements, while the fallback path sends the same prompts to a different processor, region, retention policy, abuse-monitoring regime, or training setting. The current skill asks for provider terms and API settings, but it does not require execution-time evidence for region pinning, failover destinations, model-router rules, or "no cross-border fallback" tests. This can produce an assessment that passes the nominal provider while missing the actual outage path.

**Missed variant 2: prompt cache, tracing, or observability layer stores raw prompt copies outside the primary retention policy**
```ts
await llmGateway.chat({
  model: "eu-support-assistant",
  messages,
  cache: {
    enabled: true,
    keyIncludesPrompt: true,
    region: "us-east-1",
    ttlDays: 90,
  },
  tracing: {
    exportPrompts: true,
    destination: "vendor-saas",
    redact: false,
  },
});
```

**Why it should be caught:**

The application database and LLM provider can both have acceptable retention controls while an AI gateway, prompt cache, tracing backend, evaluation dataset exporter, or observability pipeline retains the same prompt/completion data in a different region or for a longer period. The current retention section checks conversation logs, vector stores, datasets, checkpoints, and backups, but it does not explicitly require inventory and evidence for AI middleware copies such as prompt caches, LLM gateway traces, eval replay stores, or OpenTelemetry spans containing prompt attributes.

**Missed variant 3: model router selects provider by latency or cost without tenant privacy constraints**
```yaml
router:
  strategy: lowest_latency
  default_chain:
    - azure-openai-westeurope
    - bedrock-us-east-1
    - public-openai
tenant_policy:
  eu_customers:
    allowed_regions: ["eu"]
    allowed_processors: ["azure-openai-westeurope"]
```

**Why it should be caught:**

Many production LLM stacks use LiteLLM-style gateways, provider fallback, or model routers. If routing policy is optimized for availability, latency, or cost but not constrained by tenant data residency and processor commitments, the privacy posture depends on runtime routing state rather than the reviewed code path. The skill should require a "provider routing matrix" that maps tenant/data class to allowed model providers, regions, retention settings, training settings, fallback behavior, and evidence from tests or logs.

## Edge Cases

- Zero-data-retention or no-training settings may apply to only some API surfaces. Batch jobs, file uploads, image/audio endpoints, eval pipelines, and abuse-monitoring logs can have different retention behavior.
- A proxy or gateway can redact prompts before sending to the model but export the unredacted original prompt to traces or metrics.
- Multi-region endpoints and DNS/load-balancer routing can move traffic during provider incidents unless region pinning is tested, not just documented.
- A DSAR deletion flow may purge the application conversation table but leave prompt-cache entries, traces, or evaluation replay datasets intact.
- Tenant-specific privacy commitments can be violated by shared cache keys if prompt cache entries are not partitioned by tenant and data classification.

## Remediation Quality

- [x] Fix resolves the vulnerability
- [x] Fix doesn't introduce new security issues
- [x] Fix doesn't break functionality
- **Issues found:** Add a dedicated "Provider Routing, Fallback, and AI Middleware Copies" evidence gate to the skill. The gate should require:
  - an inventory of every model provider, gateway, router, cache, tracer, eval store, and telemetry exporter that can receive prompt/completion content;
  - per-tenant or per-data-class allowed providers, regions, retention periods, training settings, and fallback behavior;
  - tests or operational evidence proving that restricted data does not fall back to unapproved providers or regions;
  - evidence that prompt caches, traces, spans, and eval exports are redacted, partitioned, TTL-bound, and included in DSAR/deletion propagation.

## Comparison to Other Tools

| Tool | Catches this? | Notes |
|------|:---:|-------|
| Semgrep | Partial | Can flag SDK calls, router config keys, cache flags, or trace exporters, but cannot prove DPA/region/fallback compliance without custom rules and architecture evidence. |
| CodeQL | Partial | Can model prompt data flow into SDKs or logging sinks in application code, but usually misses runtime provider routing, SaaS gateway retention, and contract-bound data residency evidence. |
| Snyk / dependency scanners | No | Dependency scanners do not evaluate model provider routing, cross-border transfer controls, or prompt observability retention. |
| Manual privacy review / DPIA | Yes | A privacy architecture review should explicitly map processors, subprocessors, transfer regions, retention, and deletion propagation for every copy of prompt/completion data. |

## Overall Assessment

**Strengths:**

The skill already covers the core AI privacy lifecycle well: training data, prompt/completion PII, retention, model memorization, EU AI Act, and consent. It also correctly calls out third-party LLM providers, DPAs, zero-data-retention settings, vector store deletion, and output-side PII filtering.

**Needs improvement:**

The skill treats "third-party LLM provider" and "retention policy" as broad review objects, but modern LLM deployments often have a routing layer and several secondary prompt copies. Without a provider routing and AI middleware copy inventory, the assessment can miss the actual path taken during fallback, caching, tracing, eval replay, or observability export.

**Priority recommendations:**
1. Add a provider routing matrix to the output template: provider, model, region, tenant/data classes allowed, DPA/subprocessor status, training setting, retention period, fallback target, and proof source.
2. Add grep targets and review questions for LLM gateways, routers, caches, tracing, eval stores, and telemetry exporters (`litellm`, `langsmith`, `langfuse`, `helicone`, `braintrust`, `openlit`, `otel`, `prompt_cache`, `fallback`, `router`, `region`).
3. Add severity guidance: unapproved cross-region fallback for regulated personal data = High/Critical; prompt traces retaining raw PII beyond the primary retention policy = High; missing routing evidence = Medium.

## Bounty Info
- [x] I have read and agree to the [CONTRIBUTING.md](../../CONTRIBUTING.md) bounty terms
- **Preferred payment method:** PayPal or GitHub Sponsors; payment details can be provided privately after maintainer acceptance.

**Comment by Errordog2** [comment]
/claim #2246

Submitted PR #2251: https://github.com/UnitOneAI/SecuritySkills/pull/2251

Scope: adds provider routing, fallback-equivalence, and prompt-cache residency evidence gates to `skills/ai-security/ai-data-privacy/SKILL.md` for the gap described in #2246.

Validation completed:
- `git diff --check`
- Frontmatter required-field check over `skills/` and `roles/`
- `index.yaml` path existence check
- Workflow-equivalent prompt-injection pattern scan from `.github/workflows/injection-scan.yml`
- Targeted `rg` checks for the new provider-routing/cache matrix sections and `v1.0.1`

No payout address, private token, cookie, seed phrase, or API key is included in the PR or this claim comment.

**Comment by JamesJi79** [comment]
/attempt #2246

I will submit VALID GAP analysis for this review.

**Comment by JamesJi79** [comment]
/attempt #2246

## VALID GAP - ai-data-privacy: provider routing, fallback, and AI middleware copy evidence gates

### Why Valid (FP confirmed)

The issue correctly identifies a real FP scenario: a reviewer evaluating Azure OpenAI calls with EU-region-bound endpoints, PII redaction, tenant-scoped user IDs, and no-training settings could still flag the code as "PII sent to third-party LLM" because the current ai-data-privacy skill does not provide a structured evidence gate for compliant regional processor routing. The benign code path - where the provider is contractually bound, region-pinned, data is redacted pre-transmission, and tenant isolation is enforced via scoped user IDs - is indistinguishable from an unapproved public fallback under the current skill template. This is a genuine gap because it forces reviewers to either give a pass without structural evidence or issue a false positive flag that cannot be resolved through the current gate structure.

### Coverage Gaps

**Gap 1 - Uncontrolled provider fallback silently crosses approved processor/region boundary** - The issue shows a primary Azure OpenAI EU path that falls back to a direct OpenAI API key provider with different DPA terms, region, and training settings. The current skill checks provider terms and API settings at the nominal provider level but does not require execution-time evidence for failover destinations, model-router rules, or cross-border fallback tests. An assessment can pass on the primary provider while the actual outage path violates data residency and processor constraints. **Need gate**: Provider routing matrix that maps every possible provider path, fallback target, region, and tenant/data-class constraint with tests or operational logs proving restricted data does not reach unapproved endpoints.

**Gap 2 - Prompt cache, tracing, and observability layers store raw prompt copies outside primary retention policy** - The issue demonstrates that an AI gateway with caching (key-includes-prompt, different region, 90-day TTL) and tracing (prompts exported unredacted to a vendor SaaS) creates data copies outside the primary LLM provider retention scope. The current retention section covers conversation logs, vector stores, datasets, checkpoints, and backups but does not explicitly require inventory and evidence for AI middleware copies - prompt caches, LLM gateway traces, eval replay stores, or OTel spans containing prompt attributes. **Need gate**: Complete middleware copy inventory including hash-key provenance, cross-region cache propagation, trace redaction status, and DSAR deletion propagation for every secondary copy.

**Gap 3 - Model router selects provider by latency or cost without tenant privacy constraints** - The issue shows a YAML router config where strategy is lowest_latency with fallback chain from Azure EU to Bedrock US to public OpenAI, and a separate tenant policy that only permits Azure EU for EU customers. If routing policy is optimized for availability or cost rather than being constrained by tenant data residency commitments, the actual privacy posture depends on runtime routing state rather than the reviewed code. The skill currently asks for general provider and retention sections but lacks a structured provider routing matrix that maps tenant/data class to allowed providers, regions, fallback behavior, and evidence from integration tests or traffic logs. **Need gate**: Per-tenant or per-data-class provider routing matrix with testable region pinning, fallback constraints, and runtime evidence that routing decisions preserve privacy commitments.

### Suggested Gates

1. **AI-DP-RT-01 - Provider Routing Matrix**: Require an inventory of every model provider, LLM gateway, router, cache, tracer, eval store, and telemetry exporter that can receive prompt/completion content. Each entry must document DPA/subprocessor status, data residency region, training-data setting (opt-in/opt-out/none), retention period, and fallback targets. Evidence source may include IaC configuration, CI/CD test assertions, or runtime traffic logs.

2. **AI-DP-FB-01 - Cross-Border Fallback Test**: Require integration tests or staging traffic logs proving that restricted data classes (EU PII, HIPAA ePHI, PCI) do not reach fallback providers or regions outside the approved set. Tests should cover provider outage scenarios, DNS failover, and multi-region endpoint routing.

3. **AI-DP-CA-01 - Middleware Copy Inventory**: Require explicit enumeration of all prompt/completion copies outside the primary LLM provider, including cache entries (with key provenance showing whether raw prompts are used), trace/span exports (with redaction status), eval replay stores, abuse-monitoring logs, and OTel exporters. Each copy must document region, TTL, redaction applied, and whether it is included in DSAR/deletion propagation workflows.
